### DISCUSSION TITLE
Tree sitter integration

### DISCUSSION THREAD

**Discussion by aciceri** [meow-edit/meow#391] [Ideas]
Hey, after having tried Helix I realized that perhaps Vim/Evil is not the best modal editing we can get, then I found meow (days ago, so I'm still a noob) and it looks great! I liked Helix, it has really good defaults, but it doesn't give you the degree of customization of Emacs.

I'm not an Emacs (lisp) expert and also my Tree sitter knowledge is limited to what I've read in their doc (and for some languages I use syntax highlighting provided by it) so please take my proposal with a grain of salt. Would it be possible to register some tree sitter "nodes" (not sure how to call them, by the way whatever [queries](https://emacs-tree-sitter.github.io/syntax-highlighting/queries/) return) as Meow things?

If you believe this is possible, would it be easy? And especially, would it make sense or would it be better if this was managed in another package and/or Emacs mode?

**Comment by DogLooksGood** [comment]
It's easy to support, query the node from tree-sitter is easy.  But meow can't integrate it as the default, since meow is designed to be dependency-free. Maybe we can do it in a separated package.

**Reply by aciceri** [reply]
Well `tree-sitter` is integrated in Emacs 29, it wouldn't be a dependency :)

By the way I think it lacks many grammars (and I use [elisp-tree-sitter](https://github.com/emacs-tree-sitter/elisp-tree-sitter) myself) and I agree that probably it's better doing this as an external package.

If I find time I would like trying to experiment this, also as an opportunity to dive in `tree-sitter` and improve my Emacs knowledge. If you have any suggestions/directions to follow please tell me.

**Reply by thiskappaisgrey** [reply]
I've been playing around with Tree-Sitter myself, and was also looking to integrate it with Meow when I have time. From my understanding, it shouldn't be too  hard to integrate Meow using Emacs-29's Tree Sitter (the reason why it doesn't support many languages is that, you'd have  to install the grammars yourself, but the API has mechanisms to do that).

**Reply by quarkw** [reply]
I think this comment is relevant
https://old.reddit.com/r/emacs/comments/10ufbx4/a_consistent_structural_editing_interface/j7bxx1y/?context=3
I think the idea is there could be a separate package that functions as an interface between the treesitter API and the native emacs `things` API. Then in meow, you could just add support for the `things` implemented by treesitter, which I think would be really easy.

**Reply by thiskappaisgrey** [reply]
[what I have so far](https://github.com/thiskappaisgrey/meow-ts) (link to the repo)


this is what I have so far. I also tried to add support for `meow-ts--next-defun` to make it kind of work like `meow-next-word`. 
I plan to add more tree-sitter objects if I have time, but so  far, meow is pretty expandable to `things` defined by external packages (except for the fact that `meow--select-expandable-p` is hardcoded, but that's an easy fix).

**Reply by DogLooksGood** [reply]
Well, `thing` selections are not expandable by design.

**Reply by thiskappaisgrey** [reply]
Is there a reason why they shouldn't be? It would make it easier for other libraries to extend meow's `thing` s and have them work like they do with words and such (not just `tree-sitter` objects but maybe also other text objects as well, like `org-mode` objects as well or something like that), without having to modify meow.

**Reply by DogLooksGood** [reply]
Well, those things you usually select and expand are all directly supported. Except paragraphs and sentences. With current implementation, you can't do expand on `thing` selection.

**Comment by DogLooksGood** [comment]
It's easy to support, query the node from tree-sitter is easy.  But meow can't integrate it as the default, since meow is designed to be dependency-free. Maybe we can do it in a separated package.

**Comment by skissue** [comment]
Hey everyone, this discussion is pretty old, but I was looking for this functionality recently, so I decided to write a package myself. You can find it [here](https://github.com/skissue/meow-tree-sitter). It's still pretty rough around the edges, but I've been using it for a bit and it's working, so I'd love some more feedback :slightly_smiling_face:. A lot of the functionality is ported from [evil-textobj-tree-sitter](https://github.com/meain/evil-textobj-tree-sitter).

**Reply by aciceri** [reply]
Just tried it out (with the Python and Nix grammars) and it seems to work great :muscle:
I couldn't wait for this! Will give you feedback once I've it.

**Reply by eshrh** [reply]
Cool! I'll be using this.
